workflow/function-cacher/index.js

```diff
@@ -0,0 +1,72 @@
+math.config({
+  number: 'BigNumber',
+});
+
+const deepEqual = (firstObject, secondObject) => {
+  if (firstObject === secondObject) {
+    return true;
+  } else if (
+    typeof firstObject === 'object' &&
+    firstObject !== null &&
+    typeof secondObject === 'object' &&
+    secondObject !== null
+  ) {
+    if (Object.keys(firstObject).length !== Object.keys(secondObject).length) {
+      return false;
+    }
+
+    for (let key in firstObject) {
+      if (secondObject.hasOwnProperty(key)) {
+        if (!deepEqual(firstObject[key], secondObject[key])) {
+          return false;
+        }
+      } else {
+        return false;
+      }
+    }
+
+    return true;
+  } else {
+    return false;
+  }
+};
+
+class Cacher {
+  constructor() {}
+
+  withCache(callback) {
+    let cache = {};
+
+    return (...arg) => {
+      for (let key in cache) {
+        if (deepEqual(JSON.parse(JSON.stringify(arg)), JSON.parse(key))) {
+          return cache[key];
+        }
+      }
+
+      return (cache[JSON.stringify(arg)] = callback(...arg));
+    };
+  }
+}
+
+const cacher = new Cacher();
+
+const calculateFactorial = a => {
+  return math.evaluate(`${a}!`);
```

Comment:
For info:
as well math js has it's own `math.factorial(n)` function
the only difference, you need not to set config for operate with `BigNumber`, but wrap factorial argument into `math.bignumber`.